[Merged by Bors] - feat(data/nat/cast): generalize to `fun_like`

ericrbg opened the pull request:
---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

This will be followed up with a PR generalising the `ext` results below, but they still have issues that need to be fixed.

- [x] depends on: lean#657

ericrbg committed:
feat(data/nat/cast): slightly generalize

ericrbg committed:
finish things but unsure

ericrbg committed:
fix1

ericrbg committed:
fix2

ericrbg committed:
fix3

ericrbg committed:
fix3: the big fix

ericrbg committed:
the big fix 4: electric boogaloo

this is not gonna be a fun PR

jcommelin commented on src/algebra/ne_zero.lean:
```diff
@@ -55,8 +55,8 @@ lemma of_injective' {r : R} [has_zero R] [h : ne_zero r] [has_zero M] [zero_hom_
 lemma of_injective [non_assoc_semiring M] [non_assoc_semiring R] [h : ne_zero (n : R)]
-  {f : R →+* M} (hf : function.injective f) : ne_zero (n : M) :=
- ⟨λ h, (ne_zero.ne' n R) $ hf $ by simpa⟩
+  [ring_hom_class F R M] {f : F} (hf : function.injective f) : ne_zero (n : M) :=
```
I think I didn't realise last time that `(n : R)` and `(n : M)` are casting a `n : ℕ` to other rings. I understand that this is a useful lemma, but maybe it should be called `nat_of_injective`.
And `of_injective` just says that if you have an injective function `R → M` that preserves zero (i.e. a `zero_hom_class`) then it also preserves `ne_zero`.

ericrbg replied:
Yeah, sounds good;will rename tomorrow when at PC

Vierkantor replied:
Could you address this comment?

jcommelin commented on src/data/mv_polynomial/basic.lean:
```diff
@@ -1082,6 +1082,9 @@ lemma aeval_eq_eval₂_hom (p : mv_polynomial σ R) :
 
 @[simp] lemma aeval_C (r : R) : aeval f (C r) = algebra_map R S₁ r := eval₂_C _ _ _
 
+-- `simp` doesn't much like `map_nat_cast`.
```
What does this comment mean? We certainly want `simp` to trigger `map_nat_cast` simplifications. If that is not possible for certain reasons (out_params, or something like that??) then I fear this is a regression.

ericrbg replied:
Most of the time, it works (even within simp), but I had to fix a couple that didn't; this version is harder to elaborate, so in very meta-variably states (like term proofs) this new map_nat_cast can just fail to trigger. However, as a stop-gap we can just put the old versions back as a corollary of this lemma if it starts being an issue. This comment was made after much frustration...

ericrbg replied:
This is now no longer true after the kernel improvements.

Vierkantor commented on src/data/mv_polynomial/basic.lean:
```diff
@@ -1082,6 +1082,8 @@ lemma aeval_eq_eval₂_hom (p : mv_polynomial σ R) :
 
 @[simp] lemma aeval_C (r : R) : aeval f (C r) = algebra_map R S₁ r := eval₂_C _ _ _
 
+@[simp] lemma aeval_map_nat_cast (n : ℕ) : aeval f (n : mv_polynomial σ R) = n := map_nat_cast _ n
```
Should this just be:
```suggestion
@[simp] lemma aeval_nat_cast (n : ℕ) : aeval f (n : mv_polynomial σ R) = n := map_nat_cast _ n
```

ericrbg replied:
For future reference, this was no longer needed after the fixes, either.

Vierkantor commented on src/data/nat/cast.lean:
```diff
@@ -262,73 +262,88 @@ by induction n; simp *
+  apply @ext_nat' R, -- why does this need to be like this? even `ext_nat f g` doesn't work
+  simp only [map_one]
+end
 
```
[Looks like you uncovered a bug in the elaborator!](https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/Instance.20parameters.20depending.20on.20.60out_params.60) Until the fix of [lean#657](https://github.com/leanprover-community/lean/pull/657) gets released, we can help Lean along a bit by writing:
```suggestion
ext_nat' f g $ show f 1 = g 1, by simp only [map_one]
```

Vierkantor replied:
Most of the other cases should be fixable analogously, by putting in `show` to remind Lean that it knows the correct type all along.

ericrbg replied:
I think it's definitely better to wait for that fix to be merged, so that we can see how big the elaboration changes are. I'm going to remove `awaiting_review` until then. Thanks so much for the core lean fix!

Vierkantor replied:
Can this now be reverted to:
```suggestion
ext_nat' f g (by simp only [map_one])
```
If not, I'd still prefer
```suggestion
ext_nat' f g $ show f 1 = g 1, by simp only [map_one]
```

Vierkantor reviewed (commented):
Sorry for the long wait, hopefully the fix to Lean makes up for it!

ericrbg committed:
Merge branch 'master' into ericrbg/nat_cast1

ericrbg committed:
(temporarily) protect `map_nat_cast`

this will be fixed in #11161, which turns `map` into a `ring_hom`
(currently `map_ring_hom`): in this case, the normal map_nat_cast will apply

ericrbg committed:
generalize the lemma I meant to generalize v2

ericrbg committed:
defeq check not as deep?

ericrbg committed:
generalize `ne_zero`

ericrbg committed:
fixlint

Vierkantor commented on src/data/complex/is_R_or_C.lean:
```diff
@@ -362,7 +362,7 @@ by simp only [←sqrt_norm_sq_eq_norm, norm_sq_conj]
 
 @[simp, norm_cast, priority 900] theorem of_real_nat_cast (n : ℕ) : ((n : ℝ) : K) = n :=
-of_real_hom.map_nat_cast n
+show (algebra_map ℝ K) n = n, from map_nat_cast of_real_hom n
```
Can we now revert this to:
```suggestion
map_nat_cast of_real_hom n
```

ericrbg replied:
this was the only one that didn't typecheck :/ I'm not sure why, I was thinking of asking Mario on zulip as he was just talking about non-transitive defeq earlier today

Vierkantor replied:
Aha, this looks like it's same issue as in #11238: sometimes Lean can't match a concrete instance such as `ring_hom.has_coe_to_fun` with a general instance `ring_hom_class.to_has_coe_to_fun` if the types don't quite give enough information yet to infer a `ring_hom_class` instance. I can't find an easy tweak in the defeq check to make this work, so the `show` will have to stay.

Vierkantor reviewed (approved):
Thanks for the updates! I still see a few small awkwardnesses, if you could address those everything looks good to me.

bors d+

ericrbg committed:
reviewer comments
